[WIP]Graphql update (#23414)
MystenLabs/sui: docs/content/concepts/graphql-rpc.mdx

---

> title: GraphQL for Sui RPC (Beta)

Reviewer: It would be good to add a info or warn admonition at the top of this doc (before the "Headers" section) to ask developers to migrate away from the GraphQL alpha endpoints https://sui-mainnet.mystenlabs.com/graphql and https://sui-testnet.mystenlabs.com/graphql to the new GraphQL beta endpoints of https://graphql.mainnet.sui.io/graphql and https://graphql.testnet.sui.io/graphql due to the schema changes. We will turn off the alpha endpoints soon.

:::

:::caution

Switch any apps that still use the GraphQL Alpha endpoints (`https://sui-mainnet.mystenlabs.com/graphql` and `https://sui-testnet.mystenlabs.com/graphql`) to the GraphQL Beta endpoints as soon as possible to avoid disruption of service.

- GraphQL Beta Mainnet: https://graphql.mainnet.sui.io/graphql

> - GraphQL Beta Testnet: https://graphql.testnet.sui.io/graphql

Reviewer 2: Chris very kindly set-up the devnet URL as well:

```suggestion
- GraphQL Beta Testnet: https://graphql.testnet.sui.io/graphql
- GraphQL Beta Devnet: https://graphql.devnet.sui.io/graphql
```

Reviewer 2: And I guess we should mention `https://sui-devnet.mystenlabs.com/graphql` as one of the alpha URLs, in the paragraph above.

Reviewer 2: The `curl` call in the "Headers" section below also needs to be updated with the new URL, and the output regenerated with the beta service -- there may be other similar cases as well.

There are also queries embedded in those `curl` calls that need to be tweaked. E.g. in the "Variables" section, the query should look like this:

```sh
curl -X POST https://graphql.testnet.sui.io/graphql \
    --header 'Content-Type: application/json'       \
    --data '{
        "query": "query ($id: Int) { epoch(epochId: $id) { referenceGasPrice } }",
        "variables": { "id": 100 }
    }'
```

Reviewer 2: Ditto for the Guide page

#### Available range

> The GraphQL service does not support consistent pagination for arbitrarily old cursors. A cursor can grow stale, if the checkpoint it is from is no longer in the **available range**. Data outside of the available range can also be pruned away. This range might be different depending on the specific query and filters used. You can query the upper- and lower-bounds of that range for your query as follows:

Reviewer 2: This passage is a bit too specific to retention for the live object set queries, we should generalise it (and maybe rename the section to "Retention" as well). Feel free to rephrase, but this is what I came up with:

```suggestion
Different queries are subject to their own retention policies. For example, Live object set queries only work for recent checkpoints (measured in minutes or hours), and transaction pagination may only work going back to a certain checkpoint (measured in weeks or months) with different retentions supported based on the filters applied. Data outside these ranges will be pruned away. You can query the upper- and lower-bounds of that range for your query as follows:
```

```graphql
{

>   retention(type: "Query", field: "transactions", filter: "affectedAddress") {
>     first { sequenceNumber }
>     last { sequenceNumber }
>   }

Reviewer 2: ```suggestion
  serviceConfig {
    retention(type: "Query", field: "transactions", filter: "affectedAddress") {
      first { sequenceNumber }
      last { sequenceNumber }
    }
  }
```
